### DISCUSSION TITLE
Try to understand the gradient of testing node

### DISCUSSION THREAD

**Discussion by juanshu30** [pyg-team/pytorch_geometric#4556] [General]
In the example captum_explainability.py, I didn't quite understand the code(node level explanation):

captum_model = to_captum(model, mask_type='node', output_idx=output_idx)
ig = IntegratedGradients(captum_model)
ig_attr_node = ig.attribute(data.x.unsqueeze(0), target=target,
                            additional_forward_args=(data.edge_index),
                            internal_batch_size=1)

If the target node is a test node, then how do we get the gradient of the test node feature. I guess it should be the gradient of loss with respect to the test node feature, then I am just wondering this loss is the training loss or the loss of all the nodes(including training node, val node and testing node)?

Thanks!

**Comment by rusty1s** [comment]
`ig_attr_node` will return the importance of node features across all nodes and all features. If you only care about the importance of features of the test node, you can do `ig_attr_node.squeeze(0)[output_idx]`.

**Reply by juanshu30** [reply]
> `ig_attr_node` will return the importance of node features across all nodes and all features. If you only care about the importance of features of the test node, you can do `ig_attr_node.squeeze(0)[output_idx]`.

Thanks for your reply. I mean when calculating the integrated gradient of a single node feature (gradient of loss w.r.t feature), the loss is the node-wise(a single node) loss or the training loss on all training nodes?  Thanks!

**Reply by rusty1s** [reply]
It is based on all the training nodes that are reachable within the k-hop neighborhood around your node that you want to explain (due to the way message passing works and which nodes are able to determine the given node label).

**Reply by juanshu30** [reply]
Thanks! You mean we will not use the label of this test node when we explain it? We just use the loss calculated by its k-hop training nodes, and then take derivative of this loss with respect to the test node features?

How about edges? If an edge has two test node on it end, how it is explained? 

Thanks!

**Reply by rusty1s** [reply]
For explaining, we will use the label/output of the node we want to predict (this is the `target` in the code snippet above). The same holds true for edges, where we have a `target`/output per edge. For more information about how `IntegratedGradients` work, please also have a look at [this tutorial](https://captum.ai/docs/extension/integrated_gradients).
